metrics/sinks/elasticsearch/ElasticsearchSink.cpp

```diff
@@ -0,0 +1,251 @@
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "httpclient.hpp"
+#include "ElasticsearchSink.hpp"
+#include "IMeasurement.hpp"
+#include "IMetricSet.hpp"
+#include <map>
+#include <algorithm>
+#include "jstring.hpp"
+
+
+using namespace hpccMetrics;
+
+extern "C" MetricSink* getMetricSinkInstance(const std::map<std::string, std::string> &parms)
+{
+    MetricSink *pSink = new ElasticsearchSink(parms);
+    return pSink;
+}
+
+
+
+ElasticsearchSink::ElasticsearchSink(const std::map<std::string, std::string> &parms) :
+    MetricSink{parms}
+{
+    protocol.append("http");
+    port.append("9200");
+    domain.append("localhost");   // this will be a config parameter
+}
+
+
+void ElasticsearchSink::send(const std::vector<std::shared_ptr<IMeasurement>> &values, const std::string &setName)
+{
+
+    auto it = metricSets.find(setName);
+    if (it != metricSets.end())
+    {
+        //
+        // Do this every time in case the index name changes because of something like a day change.
+        MetricSetInfo &curMetricSetInfo = metricSetInfo[setName];
+        initializeIndex(it->second);
+
+        //
+        // Build the JSON payload directly
+        std::string payload;
+        payload.append("{");
+        for (auto const &pMeas : values)
+        {
+            payload.append("\"").append(pMeas->getName()).append("\" : ");
+            payload.append("\"").append(pMeas->toString()).append("\", ");
+        }
+        payload.erase(payload.find_last_of(','));  // take off the last ',';
+        payload.append("}");
+
+        StringBuffer content(payload.c_str());
+        Owned<IHttpClientContext> httpctx = getHttpClientContext();
+        StringBuffer url;
+        url.append(protocol.c_str()).append("://").append(domain.c_str()).append(":").append(port.c_str());
+        url.append("/").append(curMetricSetInfo.lastIndexName.c_str());
+        url.append("/_doc/");
+        Owned<IHttpClient> httpClient = httpctx->createHttpClient(nullptr, url);
+
+        StringBuffer resp, status;
+        int ret = httpClient->sendRequest("POST", "application/json", content, resp, status);
+        if (ret == 0)
+        {
+            int statusCode = atoi(status.str());
+        }
+    }
+    // else a throw here, or is it trusted
+}
+
+
+// Index pattern from filebeat:  index: "filebeat-%{[agent.version]}-%{+yyyy.MM.dd}"
+
+void ElasticsearchSink::init()
+{
+    // nothing to do
```

Comment:
This is probably a good thing to revisit once we have metrics being reported. It could be something that gets logged. But for now, I'm not really sure what it would look like.